flow/diff_context.h

```diff
@@ -0,0 +1,235 @@
+#ifndef FLUTTER_FLOW_DIFF_CONTEXT_H_
+#define FLUTTER_FLOW_DIFF_CONTEXT_H_
+
+#include <map>
+#include <vector>
+#include "flutter/fml/logging.h"
+#include "flutter/fml/macros.h"
+#include "third_party/skia/include/core/SkMatrix.h"
+#include "third_party/skia/include/core/SkRect.h"
+
+namespace flutter {
+
+#ifndef NDEBUG
+#define FLUTTER_ENABLE_DIFF_CONTEXT
+#endif
+
+#ifdef FLUTTER_ENABLE_DIFF_CONTEXT
+
+class Layer;
+
+struct Damage {
+  // This is the damage between current and previous frame;
+  // If embedder supports partial update, this is the region that needs to be
+  // swapped.
+  SkIRect surface_damage;
+
+  // Reflects actual change to target framebuffer; This is surface_damage +
+  // damage previously acumulated for target framebuffer.
+  // All drawing will be clipped to this region. Knowing the affected area
+  // upfront may be useful for tile based GPUs.
+  SkIRect buffer_damage;
+};
+
+// Every layer has a PaintRegion that covers screen area where the layer subtree
+// painted anything.
+//
+// The area is used when adding damage of removed or dirty later.
+//
+// Because there is a PaintRegion for each layer, it must be able to represent
+// the area with minimal overhead. This is accomplished by having one
+// vector<SkRect> shared between all paint regions, and each paint region
+// keeping begin and end index of rects relevant to particular subtree.
+//
+// All rects are in screen coordinates.
+class PaintRegion {
+ public:
+  PaintRegion() {}
+  PaintRegion(std::shared_ptr<std::vector<SkRect>> rects,
+              size_t from,
+              size_t to,
+              bool has_readback)
+      : rects_(rects), from_(from), to_(to), has_readback_(has_readback) {}
+
+  std::vector<SkRect>::const_iterator begin() const {
+    FML_DCHECK(is_valid());
+    return rects_->begin() + from_;
+  }
+
+  std::vector<SkRect>::const_iterator end() const {
+    FML_DCHECK(is_valid());
+    return rects_->begin() + to_;
+  }
+
+  // Compute bounds for this region
+  SkRect GetBounds() const;
```

Comment:
For methods that are named `Get*` i'd assume that the response is a O(1) operation, this and `GetDamage` at least seem to be doing some work, what do you think of renaming them to `Compute*`?